### PR TITLE
Docker compose setup for dev and integration tests

### PR DESCRIPTION
**Pull request by mihaibivol**
## Howto:

Get the latest Docker setup.

* If you are using Mac, please build a simple box with `docker-engine` above `1.10` and `docker-compose` above `1.6.0`.

Make sure you can run docker without sudo.
* `id $USER`
If you are not in the `docker` group, run the following command and then restart `docker`. If this doesn't work, just restart your machine :)
*  `sudo usermod -a -G docker $USER`

Get docker-compose:
```shell
sudo pip install docker-compose
```

* Add `DOCKER_DATA` env variable in your `.bashrc` or `.zshrc`. In this directory you will have all the persistent data between Docker runs.
```
export DOCKER_DATA=~/inspirehep_docker_data/
```
By default the virtualenv and everything else will be kept on `/tmp` and they will be available only until the next reboot.

* Install a host persistent venv and build assets
```shell
docker-compose pull
docker-compose -f docker-compose.deps.yml run --rm pip
docker-compose -f docker-compose.deps.yml run --rm assets
```

* Run the service locally
```shell
docker-compose up
```
Go to `localhost:5000`

* Populate database

```shell
docker-compose run --rm web scripts/recreate_records
```

* Run tests in an __isolated__ environment:

```shell
docker-compose -f docker-compose.test.yml run --rm test
docker-compose -f docker-compose.test.yml kill
docker-compose -f docker-compose.test.yml rm
```

__The kill command is mandatory, otherwise you will use the test DB instead of the correct one in dev.__

I don't know how to kill all the other services just after the run command exited.

## Extra useful tips

* Run a random shell

After `docker-compose up` just run
```
docker-compose run --rm web inspirehep shell
```

* Reload code in a worker

With `docker-compose up` just run:
```
docker-compose restart worker
```

* Quick and safe reindex

With `docker-compose up` just run:
```
docker-compose restart worker && docker-compose run --rm web scripts/recreate_records
```
## TODOs:

* [x] Fix indexing
* [x] Add test compose file
* [x] Make data persistent for dev

## Future Stuff:

* Use build arg to actually create a `staging` or `prod` image.

### CODE REVIEW

**Comment by mihaibivol** [comment]
@inspirehep/inspire-dev Feel free to comment on how these commands can be further simplified. e.g. I'm not sure how to drop the lengthy `docker-compose -f docker-compose.build.yml`

File: services.yml [hunk 1]
```diff
@@ -0,0 +1,19 @@
+version: '2'
+
+services:
+  base:
+    build: .
+    image: inspirehep/inspire-next
+  indexer:
+    image: elasticsearch
+    command: elasticsearch -Dcluster.name="inspire"
```
Comment (kaplun) [inline services.yml]: Is there some preparatory steps where one can install icu and reindexing plugin?

File: docker-compose.deps.yml [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+     volumes_from:
+       - static
+   assets:
+     extends:
+       file: services.yml
+       service: base
+     command: scripts/clean_assets
+     volumes_from:
+       - static
+  static:
```
Comment (jacquerie) [inline docker-compose.deps.yml]: Here I get an error about indentation
```shell
ERROR: yaml.parser.ParserError: while parsing a block mapping
  in "./docker-compose.deps.yml", line 1, column 1
expected <block end>, but found '<block mapping start>'
  in "./docker-compose.deps.yml", line 18, column 3
```
which is easy to fix, but I wonder why your docker wasn't complaining.
```shell
$ docker --version
Docker version 1.10.3, build 20f81dd
```

**Comment by jacquerie** [comment]
I almost got this working. One problem under Mac OS is that ports aren't forwarded, so one has to type `UGLY_URL:5000` to access the web app, or set up port forwarding in Virtualbox, or forward all ports using a local nginx (http://stackoverflow.com/questions/25327012/access-docker-from-external-machine-in-network).

**Comment by jacquerie** [comment]
Actually, one **has** to enable one of these solutions, otherwise the application will return a 404 (I have no idea why...).

Other than that, I measured that this solution runs tests 5X slower than before, but achieves better isolation and reproducibility, a tradeoff which I think we can afford (it's 12 seconds vs 1 minute). Moreover, disk usage is just 300MB, probably thanks to the `--rm` option.

So it's a :+1: for me. @kaplun, if we decide to go forward on this (it's your call anyway!) I think we should have a Monday meeting to explain the new developer workflow that @mihaibivol is building.

**Comment by kaplun** [comment]
Eheh, Indeed I'd like to understand it as well... What do you think to do when we introduce beard and the thousands of compiled dependencies?

**Comment by mihaibivol** [comment]
Currently everything that we "build" is cached in /virtualenv in the `base` service. This means that for dev at least you only have to install the requirements once and update them exactly as you would do when using a local venv. I am still not sure what's the best take for running this setup on Travis.

File: .travis.yml [hunk 3]
```diff
@@ -17,50 +17,47 @@
...
-  apt:
-    packages:
-      - nodejs
-
 notifications:
   email: false
 
+# Not needed at the moment since we use Docker.
+# If this kicks in we can try to use the cache
 sudo: false
```
Comment (jacquerie) [inline .travis.yml]: Maybe we can give up and have `sudo: required`?

Comment (mihaibivol) [inline .travis.yml]: The builds should start failing when they implement this in Docker (probably not too soon) and we'll know to update our build

File: .travis.yml [hunk 4]
```diff
@@ -17,50 +17,47 @@
...
+    - DOCKER_VERSION=1.10.1-0~trusty
 
 python:
   - "2.7"
 
 before_install:
-  - "travis_retry pip install --upgrade pip setuptools py mock pytest"
-  - "travis_retry pip install twine wheel coveralls"
-  - "npm update && npm install --silent -g node-sass clean-css uglify-js requirejs"
+  - pip install docker-compose
```
Comment (jacquerie) [inline .travis.yml]: What's the reason we don't use the linked post's solution for fixing `docker-compose` version?

File: docker-compose.test.yml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    extends:
+      service: service_base
+    command: celery worker -E -A inspirehep.celery --loglevel=INFO --purge
+    volumes_from:
+      - static
+    links:
+      - database
+      - indexer
+      - rabbitmq
+      - redis
```
Comment (jacquerie) [inline docker-compose.test.yml]: Aren't we missing `workermon`? It's seldom useful, but it's invaluable in those cases.

File: dockerfiles/docker_entrypoint.sh [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+    virtualenv /virtualenv
+    source /virtualenv/bin/activate
+    pip install --upgrade pip-accel setuptools
+else
+    source /virtualenv/bin/activate
+fi
+
+clear_pycache() {
+    find \( -name __pycache__ -o -name '*.pyc' \) -delete
+}
```
Comment (jacquerie) [inline dockerfiles/docker_entrypoint.sh]: This is just defined, not executed?

File: scripts/clear_pycache [hunk 7]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+sudo find \( -name __pycache__ -o -name '*.pyc' \) -delete
```
Comment (jacquerie) [inline scripts/clear_pycache]: As I was saying the other day, not cross-platform. On Mac OS X:
```
% find \( -name __pycache__ -o -name '*.pyc' \) -delete
usage: find [-H | -L | -P] [-EXdsx] [-f path] path ... [expression]
       find [-H | -L | -P] [-EXdsx] -f path [path ...] [expression]
```

File: scripts/recreate_records [hunk 8]
```diff
@@ -36,26 +36,6 @@ if [ ! -x "$(command -v celery)" ]; then
...
-CELERY_PIDFILE="${VIRTUAL_ENV}/worker.pid"
-if [ -f "${CELERY_PIDFILE}" ]; then
-	echo "----> Purging all pending messages"
-	celery purge -f -A inspirehep.celery --workdir="${VIRTUAL_ENV}"
-else
-	celery worker -E -A inspirehep.celery --loglevel=${CELERY_LOGLEVEL} --workdir="${VIRTUAL_ENV}" --autoreload --pidfile="${CELERY_PIDFILE}" --purge &
-	CELERY_PID=$!
-	trap 'cleanup' 0 1 2 3 6
-fi
-
```
Comment (jacquerie) [inline scripts/recreate_records]: As a general indication, but not something that should be done now, @jalavik suggests that we keep the doors open for the developers that don't want to ride the Docker train. So this script will have to be split in two cases, one run by docker, one run by non-Docker users.

File: docker-compose.test.yml (hunk 5 above)
Comment (mihaibivol) [inline docker-compose.test.yml]: can you add it and test it? I'm not sure how it exactly works

File: .travis.yml (hunk 4 above)
Comment (mihaibivol) [inline .travis.yml]: This has the same version and makes .travis.yml less verbose. Also it's the same as I did locally.

Comment (jacquerie) [inline .travis.yml]: Yeah, but isn't his approach more robust? For example, let's say that the version on pip updates to 2.0.0, which is incompatible with the Docker version we require.

File: docker-compose.test.yml (hunk 5 above)
Comment (jacquerie) [inline docker-compose.test.yml]: Sure. Its actual name is `flower`, by the way.

File: services.yml [hunk 9]
```diff
@@ -0,0 +1,19 @@
+version: '2'
+
+services:
+  base:
+    image: mihaibivol/python-base
```
Comment (jalavik) [inline services.yml]: I added you to the inspirehep dockerhub organization, so can you push these there?

**Comment by jalavik** [comment]
Awesome stuff @mihaibivol. I will give this a closer look and test.

File: docker-compose.test.yml (hunk 5 above)
Comment (jacquerie) [inline docker-compose.test.yml]: Added [here](https://github.com/jacquerie/inspire-next/commit/f321efc7fee525838108621f5687055051030043) and tested working on my machine.

**Comment by jacquerie** [comment]
If Mac OS X developers want to test the Docker way here's a Vagrantfile to help them do so: https://github.com/jacquerie/inspire-next/commit/7ddb99c8779d6374be46f9808df81040fe927a1d.
 Unfortunately I wasn't able to have docker-machine running + ES (known issue of mismatched UIDs), so for now Vagrant is the way.

**Comment by jacquerie** [comment]
...you're going to hate me a bit, but here's another commit you should cherry-pick and squash: https://github.com/jacquerie/inspire-next/commit/071ce8543b59afbffe31828887bf37d101bf18f6

This one adds the license headers to the files you introduced.

File: .travis.yml [hunk 10]
```diff
@@ -17,50 +17,56 @@
...
+  - apt-cache madison docker-engine
+  # Update Docker to 1.10.1 https://graysonkoonce.com/managing-docker-and-docker-compose-versions-on-travis-ci/.
+  - sudo apt-get -o Dpkg::Options::="--force-confnew" install -y docker-engine=${DOCKER_VERSION}
+  # Add docker-compose at the version specified in ENV.
+  - sudo rm /usr/local/bin/docker-compose
+  - curl -L https://github.com/docker/compose/releases/download/${DOCKER_COMPOSE_VERSION}/docker-compose-`uname -s`-`uname -m` > docker-compose
+  - chmod +x docker-compose
+  - sudo mv docker-compose /usr/local/bin
+  # Put coveralls
+  - pip install coveralls
```
Comment (jacquerie) [inline .travis.yml]: `travis_retry` to avoid trivial build errors?

Comment (jacquerie) [inline .travis.yml]: And maybe remove the comment and put it before everything, where it was. Or does that break stuff?

File: .travis.yml [hunk 11]
```diff
@@ -17,50 +17,56 @@
...
-  - "python setup.py test"
+  - "docker-compose -f docker-compose.test.yml run --rm test"
+
+after_script:
+  - "docker-compose -f docker-compose.test.yml kill"
+  - "docker-compose -f docker-compose.test.yml rm -f"
 
 after_success:
+  # Coverage report contains Docker paths. Replace so
+  # we can run coveralls from Travis
```
Comment (jacquerie) [inline .travis.yml]: Nit: `# Coverage report contains Docker paths. We replace them, so that we can run Coveralls.`

File: scripts/clear_pycache [hunk 12]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+sudo find . \( -name __pycache__ -o -name '*.pyc' \) | xargs rm -rf
```
Comment (jacquerie) [inline scripts/clear_pycache]: I wonder if we can be a little smarter and use `sudo` only when it's actually needed. But it's not a blocker for this PR, just a random thought.

**Comment by jacquerie** [comment]
LGTM. Cherry-pick those two commits, squash, incorporate those nits and :shipit:

**Comment by kaplun** [comment]
Guys you have my full trust on this branch :-) Don't know if @jalavik wanted to further comment...

**Comment by jacquerie** [comment]
:ship: :ship: :ship:

**Comment by jalavik** [comment]
Great stuff guys. Note that it is still optional if one wants to develop using docker. This branch does not enforce it.

**Comment by jacquerie** [comment]
BTW we kept the Vagrantfile out of this PR because we are waiting to see if Docker for Mac could be an alternative: https://blog.docker.com/2016/03/docker-for-mac-windows-beta/.

**Comment by rikirenz** [comment]
I have tried to run the command `docker-compose -f docker-compose.test.yml run --rm test` but an error is triggered. Is it my fault? am I missing something?

**Pull request merged**